[DataLoader] Change Decoder signature and add MatHandler

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57391 [DataLoader] Change Decoder signature and add MatHandler**

- Change the signature of RoutedDecoderIterDataPipe
  - Add `functional_datapipe`
  - Change handler from list to *args
  - Add `key_fn` for different methods to dispatch handlers (Default is to extract file extension from pathname)
- Change behavior of Decoder and Handler
  - Decoder takes charge of find key using `key_fn` (file extension)
  - Handlers don't need to handler extension extraction. (Remove duplicate code)
- Add MatHandler

Differential Revision: [D28151601](https://our.internmc.facebook.com/intern/diff/D28151601)

ejguan committed:
[DataLoader] Change Decoder signature and add MatHandler

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Change Decoder signature and add MatHandler"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Change Decoder signature and add MatHandler"




[ghstack-poisoned]

pmeier commented on torch/utils/data/datapipes/iter/routeddecoder.py:
```diff
@@ -1,38 +1,47 @@
 
+# Extract extension from pathname
+def _default_key_fn(pathname):
+    return re.sub(r".*[.]", "", pathname)
```
What is the advantage over `os.path.splitext(pathname)[1]`?

ejguan replied:
Advantage of `re.sub(r".*[.]", "", pathname)`:
1. Directly output extension without `.`
2. For `.temp`, it returns `temp` rather than empty string. (Not sure if empty string should be expected as a key)

Advantage of `os.path.splitext(pathname)[1]`
1. Can take pathlike object

ejguan replied:
Second thought, maybe it's not advantage for `.temp` file. It's just a hidden file without extension. 
I will change it into `os.path.splitext` then.

pmeier commented on torch/utils/data/datapipes/iter/routeddecoder.py:
```diff
@@ -1,38 +1,47 @@
-        if handlers:
-            self.decoder = Decoder(handlers)
+        self.datapipe: Iterable[Tuple[str, BufferedIOBase]] = datapipe
+        if len(handlers) > 0:
```
Why not stick with `if handlers:`? This is more pythonic.

ejguan replied:
Just want to explicitly check the asterisk argument.

pmeier commented on torch/utils/data/datapipes/iter/routeddecoder.py:
```diff
@@ -1,38 +1,47 @@
     args:
         datapipe: Iterable datapipe that provides pathname and binary stream in tuples
-        handlers: user defined decoder handlers, if None, basic and image decoder handlers will be set as default
-        length: a nominal length of the datapipe
```
Will the removal of the `length` parameter happen to all datapipes?

ejguan replied:
`length` is an optional argument for all datapipes. Only for the datapipe that can not be inferred for the length based on arguments, `length` argument is needed 
For this DataPipe, the length should remain same as the input datapipe. So, no need to have length as an argument, just like all other funcitonal datapipe such as Map.

pmeier commented on torch/utils/data/datapipes/iter/routeddecoder.py:
```diff
@@ -41,7 +50,7 @@ def __iter__(self) -> Iterator[Tuple[str, Any]]:
-            raise NotImplementedError
-        return self.length
+    def __len__(self) -> int:
+        if isinstance(self.datapipe, Sized) and len(self.datapipe) >= 0:
```
I think the second part of this check is unnecessary: if the datapipe is a `Sized` collection shouldn't its length always be `>=0`?

ejguan replied:
Emmm. I was thinking about the previous datapipe can have `-1` as the length. You are right, if the prior datapipe has -1 as length, it should raise `NotImplementedError`, so the second argument can be deleted.

pmeier commented on torch/utils/data/datapipes/utils/decoder.py:
```diff
@@ -30,6 +26,7 @@ def basichandlers(key, data):
             return None
 
     if extension in "json jsn":
+        import json
```
Why import `json` lazily? IIRC it has no system dependencies and is available in the standard library.

pmeier commented on torch/utils/data/datapipes/iter/routeddecoder.py:
```diff
@@ -1,38 +1,47 @@
-        if handlers:
-            self.decoder = Decoder(handlers)
+        self.datapipe: Iterable[Tuple[str, BufferedIOBase]] = datapipe
+        if len(handlers) > 0:
+            self.decoder = Decoder(handlers, key_fn)
```
Instead of having two branches, maybe just simply set a default value if nothing is passed:

```suggestion
        if not handlers:
            handlers = [decoder_basichandlers, decoder_imagehandler('torch')]
        self.decoder = Decoder(handlers, key_fn)
```

(this should replace the full `if` / `else`, but I can't add a suggestion for this)

pmeier commented on torch/utils/data/datapipes/utils/decoder.py:
```diff
@@ -228,29 +222,47 @@ def torch_audio(key, data):
-            self.handlers = [handler]
-        else:
-            self.handlers.append(handler)
+        self.handlers = [handler] + self.handlers
```
This avoids creating a new list over and over.

```suggestion
        self.handlers = self.handlers.insert(0, handler)
```

pmeier reviewed (commented):
Thanks @ejguan. I have a few comments inline.

ejguan committed:
Update on "[DataLoader] Change Decoder signature and add MatHandler"


- Change the signature of RoutedDecoderIterDataPipe
  - Add `functional_datapipe`
  - Change handler from list to *args
  - Add `key_fn` for different methods to dispatch handlers (Default is to extract file extension from pathname)
- Change behavior of Decoder and Handler
  - Decoder takes charge of find key using `key_fn` (file extension)
  - Handlers don't need to handler extension extraction. (Remove duplicate code)
- Add MatHandler

Differential Revision: [D28151601](https://our.internmc.facebook.com/intern/diff/D28151601)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28151601).

ejguan commented on torch/utils/data/datapipes/utils/decoder.py:
```diff
@@ -241,16 +262,16 @@ class Decoder:
+    # handler having the highest priority
+    def add_handler(self, *handler):
         if not handler:
             return
-        if not self.handlers:
-            self.handlers = [handler]
-        else:
-            self.handlers.append(handler)
+        self.handlers[0: 0] = list(handler)
```
I will leave it as it is without adding map from extension to handler for now. In the future, we should add both key-handler mapping and regex matching for decoder.

pmeier replied:
That syntax looks quite odd. Since we are not looping here, maybe:

```suggestion
        self.handlers = list(handler) + self.handlers
```

pmeier commented on torch/utils/data/datapipes/iter/routeddecoder.py:
```diff
@@ -1,47 +1,49 @@
+        handlers: Optional user defined decoder handlers. If None, basic and image decoder
+            handlers will be set as default. If multiple handles are provided, the priority
+            order follows the order of handlers (the first handler has the top priority)
+        key_fn: Function for decoder to extract key from pathname to dispatch handlers.
+            Default is set to extract file extension from pathname
```
Maybe we need to add a comment saying that if this returns anything other than the extension, the default handlers will not work and need to be also overwritten.

ejguan replied:
Make sense to me. Will add several notes to this DataPipe.

pmeier commented on torch/utils/data/datapipes/utils/decoder.py:
```diff
@@ -228,10 +222,37 @@ def torch_audio(key, data):
+################################################################
+# mat
+################################################################
+class MatHandler:
```
It seems the `ImageHandler` also has an accompanying function `imagehandler`. Maybe add a `mathandler`?

ejguan replied:
I will change `torch_video` and `torch_audio` to `videohandler` and `audiohandler` as well.

pmeier commented on torch/utils/data/datapipes/utils/decoder.py:
```diff
@@ -228,10 +222,37 @@ def torch_audio(key, data):
 # a sample decoder
 ################################################################
+# Extract extension from pathname
+def _default_key_fn(pathname):
```
`default_key_fn` is not an expressive name. Maybe `extract_extension`?

pmeier reviewed (approved):
Only minor things left. Otherwise LGTM!

ejguan committed:
Update on "[DataLoader] Change Decoder signature and add MatHandler"


- Change the signature of RoutedDecoderIterDataPipe
  - Add `functional_datapipe`
  - Change handler from list to *args
  - Add `key_fn` for different methods to dispatch handlers (Default is to extract file extension from pathname)
- Change behavior of Decoder and Handler
  - Decoder takes charge of find key using `key_fn` (file extension)
  - Handlers don't need to handler extension extraction. (Remove duplicate code)
- Add MatHandler

Differential Revision: [D28151601](https://our.internmc.facebook.com/intern/diff/D28151601)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28151601).
